Nodejs kafka tutorial - Run through - support for oauthbearer token (#600)
redhat-developer/app-services-guides: docs/kafka/nodejs-kafka/README.adoc

* You have a Kafka instance in {product-kafka} and the instance is in the *Ready* state. To learn how to create a Kafka instance, see {base-url}{getting-started-url-kafka}[Getting started with {product-long-kafka}].
* https://github.com/git-guides/[Git^] is installed.
* You have an IDE such as https://www.jetbrains.com/idea/download/[IntelliJ IDEA^], https://www.eclipse.org/downloads/[Eclipse^], or https://code.visualstudio.com/Download[Visual Studio Code^].

> * https://nodejs.org/en/download/[Node.js 16^] is installed.

Reviewer: @pacostas Just curious why we removed the noted limitation about node-rdkafka in addition to changing the minimum specified version of Node.js. Is that limitation no longer true?

Author: @jbyrne-redhat I changed it from 14 to 16 because on May will go EOL, although I agree keeping it to 14. I think it would be nice to add something like `Node.js 14 or later is installed.` if you think this would help.
Yes, this limitation no longer is true, node-rdkafka can run on node versions 14, 16 and 18.

Reviewer: Yes, I agree:

```
* https://nodejs.org/en/download/[Node.js 14^] or later is installed
```

> * https://nodejs.org/en/download/[Node.js 16^] is installed.

Author: ```suggestion
* https://nodejs.org/en/download/[Node.js 14^] or later is installed.
```

. In your IDE, create a new file. Save the file with the name `rhoas.env`, at the root level of the `reactive-example` directory for the cloned repository.

> . In the rhoas.env file, set the SASL connection mechanism and the Kafka instance client credentials as shown in the following configuration. Replace the values with your own credential information. {product-kafka} supports the SASL/OAUTHBEARER mechanism for authentication, which is the recommended authentication mechanism to use.

Reviewer: Suggest a slight rewording here:

```
. In the `rhoas.env` file, set the SASL authentication mechanism and the Kafka instance client credentials as shown in the following configuration. Replace the values with your own credential information. The configuration uses the SASL/OAUTHBEARER authentication mechanism, which is the recommended authentication mechanism to use in {product-kafka}.
```

Author: @jbyrne-redhat can you approve?

Reviewer: :+1:

+

> .Setting environment variables in the rhoas.env file

Author: ```suggestion
.Setting environment variables in the rhoas.env file [source,subs="+attributes,+quotes"] ---- KAFKA_HOST=__<bootstrap_server>__ RHOAS_SERVICE_ACCOUNT_CLIENT_ID=__<client_id>__ RHOAS_SERVICE_ACCOUNT_CLIENT_SECRET=__<client_secret>__ KAFKA_SASL_MECHANISM=plain KAFKA_SASL_MECHANISM=oauthbearer RHOAS_TOKEN_ENDPOINT_URL={sso-token-url}
```

> [source,subs="+quotes"]

Author: ```suggestion
[source,subs="+attributes,+quotes"]
```

----
KAFKA_HOST=__<bootstrap_server>__
RHOAS_SERVICE_ACCOUNT_CLIENT_ID=__<client_id>__
RHOAS_SERVICE_ACCOUNT_CLIENT_SECRET=__<client_secret>__
KAFKA_SASL_MECHANISM=oauthbearer

> RHOAS_TOKEN_ENDPOINT_URL=https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token

Reviewer: There is an available attribute for the token endpoint URL. Therefore:

```
.Setting environment variables in the rhoas.env file
[source,subs="+attributes,+quotes"]
----
KAFKA_HOST=__<bootstrap_server>__
RHOAS_SERVICE_ACCOUNT_CLIENT_ID=__<client_id>__
RHOAS_SERVICE_ACCOUNT_CLIENT_SECRET=__<client_secret>__
KAFKA_SASL_MECHANISM=plain
KAFKA_SASL_MECHANISM=oauthbearer
RHOAS_TOKEN_ENDPOINT_URL={sso-token-url}
```

Reviewer: As part of the change suggested above, observe that we updated the `subs` element at the head of the code block to also include attributes.

> RHOAS_TOKEN_ENDPOINT_URL=https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token

Author: ```suggestion
RHOAS_TOKEN_ENDPOINT_URL={sso-token-url}
```

endif::[]
+
In this case, observe that the Node.js application uses the SASL/PLAIN authentication method (that is, the value of `KAFKA_SASL_MECHANISM` is set to `plain`). This means that the application uses only the client ID and client secret to authenticate with the Kafka instance. The application doesn't require an authentication token.

> +

Reviewer 2: Could we get rid of the instruction to use PLAIN altogether here? So then we could add something like this: In the `rhoas.env` file, set the SASL connection mechanism and the Kafka instance client credentials as shown in the following configuration. Replace the values with your own credential information. {product-kafka} supports the SASL/OAUTHBEARER mechanism for authentication, which is the recommended authentication mechanism to use.

KAFKA_HOST=<bootstrap_server>
RHOAS_SERVICE_ACCOUNT_CLIENT_ID=<client_id>
RHOAS_SERVICE_ACCOUNT_CLIENT_SECRET=<client_secret>
KAFKA_SASL_MECHANISM=oauthbearer
RHOAS_TOKEN_ENDPOINT_URL=https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token

Reviewer 3: For the tutorial I'm happy to only mention OAUTHBEARER, even though the code will support both.

Author: @AedinC Is it ok to push it in a commit ? and have a review on that too? Practically i'll copy paste the above text you wrote and replace the `plain` instructions.
```
Could we get rid of the instruction to use PLAIN altogether here? So then we could add something like this: In the rhoas.env file, set the SASL connection mechanism and the Kafka instance client credentials as shown in the following configuration. Replace the values with your own credential information. {product-kafka} supports the SASL/OAUTHBEARER mechanism for authentication, which is the recommended authentication mechanism to use.

KAFKA_HOST=<bootstrap_server>
RHOAS_SERVICE_ACCOUNT_CLIENT_ID=<client_id>
RHOAS_SERVICE_ACCOUNT_CLIENT_SECRET=<client_secret>
KAFKA_SASL_MECHANISM=oauthbearer
RHOAS_TOKEN_ENDPOINT_URL=https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token
```

Reviewer: @pacostas Aedin is on PTO this week, so I'll do another review of the updates for you :+1: . I actually worked with @mhdawson on the original version of this quick start.

Author: Its a very interesting tutorial and a hot (not kafka) topic these days :p  
@jbyrne-redhat Here is the commit I just pushed, with the text pasted from Aedin https://github.com/redhat-developer/app-services-guides/pull/600/commits/73c2264fc704206b5568ffcc016a4d014f77b9ba

* Verify that the `countries` topic appears on the *Topics* page.
endif::[]

> [id="configuring-access-on-kafka-topic_{context}"]

Reviewer: @pacostas Thanks so much for adding this detailed procedure on setting access permissions for the `countries` topic.

With that said, I'm not sure this is _quite_ the approach we want to follow here. In general, we continue to add more and more examples and guides for connecting different types of clients to OpenShift Streams. Those examples all use topics with various names and require different access permissions. It's not really scalable for us to write a tailored procedure each time for setting access permissions. It also just increases the number of procedures that we must separately maintain and verify in our documentation set.

We have a [detailed guide](https://access.redhat.com/documentation/en-us/red_hat_openshift_streams_for_apache_kafka/1/guide/2f4bf7cf-5de2-4254-8274-6bf71673f407#_6b9cf9a6-5bbf-43ab-8ee6-b9f33630bf64) on setting and managing access permissions, and that should be our single point of reference as much as possible in our documentation set. That guide includes example scenarios for setting produce and consume permissions for a specified topic.

Therefore, the changes I recommend here are the following:

- We omit this dedicated procedure on configuring access permissions for the `countries` topic
- In the final `Running the Node.js example application` section, we add one more prerequisite, stating that you must have configured access permissions for the `countries` topic. That should include a link our [access management guide](https://access.redhat.com/documentation/en-us/red_hat_openshift_streams_for_apache_kafka/1/guide/2f4bf7cf-5de2-4254-8274-6bf71673f407#_6b9cf9a6-5bbf-43ab-8ee6-b9f33630bf64). I will make that suggestion on the relevant line below.

Reviewer: Based on the new prerequisite I've suggested above, we should also remove the final prerequisite in the `Configuring the Node.js example application to connect to a Kafka instance` section (line 157). That prereq is not actually needed at that point in the example anyway.

Author: I like when everything is included in a tutorial and in case I want to dig further I can follow the references. Mostly I saw it as a friction point, where despite following all the steps on the tutorial, my setup didnt work, and I had to spend time to find some extra pieces.

Although, I totally understand, duplication is not maintainable and should be avoided. I've added your suggestions under this commit https://github.com/redhat-developer/app-services-guides/pull/600/commits/5788cfae2eadac69d06b9adbe7c7d8be9d5af070

@jbyrne-redhat can you approve?

* Verify that the `countries` topic appears on the *Topics* page.
endif::[]

[id="configuring-access-on-kafka-topic_{context}"]
== Configuring access on Kafka topic

[role="_abstract"]
The application in order to produce and consume messages from Kafka topic, needs to have proper permissions. In this task you will configure access to your Kafka topic.

.Procedure
. In the {product-kafka} web console, select *Kafka Instances* and then click the name of the Kafka instance that you want to manage access.
. Select the *Access* tab.
. Click on *Manage access* button
. Select your service account and click on *next*.
+
[.screencapture]
.Overlay for managing access to Kafka Topic

> image::kafka-topic-manage-access.png[Image of overlay for managing access of kafka topic]

Reviewer 2: ```suggestion
image::kafka-topic-manage-access.png[Image of access configuration details for managing access of kafka topic]
```

+
. On the *Assign Permissions section,* on the dropdown menu select *Consume from a topic*:
+
[.screencapture]
.Overlay for managing access to Kafka Topic as a consumer

> image::consumer-permissions-for-kafka-topic.png[Image of overlay for managing cosumer's permissions]

Reviewer 2: ```suggestion
image::consumer-permissions-for-kafka-topic.png[Image of access configuration details for managing consumer permissions]
```

+

> .. On the first dropdown menu of *Topic* select `Starts with` and on the second dropdown menu, select the `countries` topic

Reviewer 2: ```suggestion
.. In the first list for *Topic* select `Starts with` and in the second list, select the `countries` topic.
```

> .. On the first dropdown menu of *Consumer group* select `Starts with` and on the second dropdown menu enter the value `consumer-test`

Reviewer 2: ```suggestion
.. In the first list of *Consumer group* select `Starts with` and for the second field, enter the value `consumer-test`.
```

. On the *Assign Permissions section,* on the dropdown menu select *Consume from a topic*:
+
[.screencapture]
.Overlay for managing access to Kafka Topic for the producer

> image::producer-permissions-for-kafka-topic.png[Image of overlay for managing producer's permissions]

Reviewer 2: ```suggestion
image::producer-permissions-for-kafka-topic.png[Image of access configuration details for managing producer permissions]
```

+

> .. On the first dropdown menu of *Topic* select `Starts with` and on the second dropdown menu, select the `countries` topic

Reviewer 2: ```suggestion
..  In the first list for *Topic* select `Starts with` and in the second list, select the `countries` topic.
```

> 

Reviewer 2: ```suggestion
. Click *Save*.
```

> 

Reviewer 2: ```suggestion
.Verification
ifdef::qs[]
* Are the accounts listed on the *Access* page?
endif::[]
ifndef::qs[]
* Verify that the accounts are listed on the *Access* page.
endif::[]
```
